docs/microbenchmark-design-guidelines.md

```diff
@@ -0,0 +1,557 @@
+
+![Rename](img/benchmark_id_rename.png)
+
+The Reporting System is used to monitor the performance changes and detect regressions. If the body of the benchmark is changed, the old name is kept and the reported time increases, it's going to be marked as regression and require a dedicated performance investigation. We want to avoid such false alarms and hence we want to reduce the number of changes in existing benchmarks to a minimum.
+
+![Regression](img/benchmark_id_regression.png)
+
+If you have some good reasons for changing the implementation of the benchmark you should remove the old one and introduce a new one with a new name.
+
+## BenchmarkDotNet
+
+BenchmarkDotNet is the benchmarking harness used in this repository. If you are new to BenchmarkDotNet, you should read [this doc](./benchmarkdotnet.md).
+
+Key things that you need to remember: 
+
+* BenchmarkDotNet **does not require the user to provide the number of iterations and invocations per iteration**, it implements a smart heuristic based on standard error and runs the benchmark until the results are not stable.
+* BenchmarkDotNet runs every benchmark in a separate process, process isolation allows avoiding side-effects. The more memory allocated by given benchmark, the bigger the difference for in-proc vs out-proc execution.
+* BenchmarkDotNet was designed to make accurate nano-benchmarks with repeatable results possible, to achieve that it does many things, including overhead calculation and subtraction (it benchmarks an empty method with the same signature and subtract the average value from results).
+* BenchmarkDotNet removes outliers by default (this repo is configured to remove only the upper outliers)
+* BenchmarkDotNet creates a type which derives from type with benchmarks. So the type with benchmarks must **not** be **sealed** and it can **not** be **static** and it has to be **public**. It also has to be a `class` (no structs support).
+
+## Setup
+
+Let's write a benchmark that measures the performance of reverting an array of 1 000 integers:
+
+```cs
+[Benchmark]
+public void Reverse()
+{
+    int[] array = Enumerable.Range(0, 1_000).ToArray();
+
+    Array.Reverse(array);
+}
+```
+
+Profile it using the [ETW Profiler](./benchmarkdotnet.md#Profiling):
+
+```cmd
+dotnet run -c Release -f netcoreapp3.0 --filter *.Reverse --profiler ETW
+```
+
+And open the produced trace file with [PerfView](https://github.com/Microsoft/perfview):
+
+![Regression](img/setup_array_reverse_profiler.png)
+
+As you can see, reversing the array took only `26.6%` of the benchmark execution time!! The rest was spent on executing the array creation logic. What does it mean? It means that the given benchmark is measuring the performance of creating and reversing the array. Not just reversing the array.
+
+This is why **the initialization logic should be always separated from the benchmark**. If possible of course.
```

Comment:
I think it is almost always possible. There are very few cases where the initialization isn't something you can remove - I'd advocate for writing docs that drive the behavior we want (which in this case is "think hard about it and strive to do the right thing.")
```suggestion
This is why **the initialization logic should be always separated from the benchmark**.
```